File: android/app/src/main/java/app/maptalk/geo/GeoPoint.kt
package app.maptalk.geo

import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * A plain latitude/longitude pair, deliberately free of any Maps or Firestore type so the
 * geo rules can be unit tested and stay comparable with the iOS implementation.
 */
data class GeoPoint(val lat: Double, val lng: Double) {

    /**
     * Great-circle distance in metres. Used to drop the false positives that geohash
     * bounds always include; both apps filter with the same formula so a thread never
     * appears on one platform and not the other.
     */
    fun distanceTo(other: GeoPoint): Double {
        val earthRadiusMetres = 6_371_008.8
        val dLat = Math.toRadians(other.lat - lat)
        val dLng = Math.toRadians(other.lng - lng)
        val a = sin(dLat / 2) * sin(dLat / 2) +
            cos(Math.toRadians(lat)) * cos(Math.toRadians(other.lat)) *
            sin(dLng / 2) * sin(dLng / 2)
        return 2 * earthRadiusMetres * asin(min(1.0, sqrt(a)))
    }

    /** Move south by [metres] (pin stays put while the camera centre shifts under a bottom sheet). */
    fun shiftedSouth(metres: Double): GeoPoint {
        val degrees = metres / 111_320.0
        return GeoPoint(lat = lat - degrees, lng = lng)
    }
}

File: ios/MapTalk/Core/GeoPoint.swift
import CoreLocation
import Foundation

/// A plain latitude/longitude pair, deliberately free of any MapKit or Firestore type so the
/// geo rules can be unit tested and stay comparable with the Android implementation.
struct GeoPoint: Equatable, Sendable {
    let lat: Double
    let lng: Double

    init(lat: Double, lng: Double) {
        self.lat = lat
        self.lng = lng
    }

    init(_ coordinate: CLLocationCoordinate2D) {
        self.lat = coordinate.latitude
        self.lng = coordinate.longitude
    }

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }

    /// Great-circle distance in metres. Android filters geohash false positives with the same
    /// formula, so both apps draw exactly the same set of bubbles.
    func distance(to other: GeoPoint) -> Double {
        let earthRadiusMeters = 6_371_008.8
        let dLat = (other.lat - lat) * .pi / 180
        let dLng = (other.lng - lng) * .pi / 180
        let a = sin(dLat / 2) * sin(dLat / 2)
            + cos(lat * .pi / 180) * cos(other.lat * .pi / 180)
            * sin(dLng / 2) * sin(dLng / 2)
        return 2 * earthRadiusMeters * asin(min(1, a.squareRoot()))
    }

    /// Move south by `meters` (pin stays put while the camera centre shifts under a bottom sheet).
    func shiftedSouth(byMeters meters: Double) -> GeoPoint {
        let degrees = meters / 111_320.0
        return GeoPoint(lat: lat - degrees, lng: lng)
    }
}
